File: 剑指offer/13-机器人的运动范围P92/Solution.java
// Author : lihongyu
// Time   : 2020-2-23
// leetcode-79 单词搜索
// 思路：回溯法
// 时间复杂度: O(m*n*m*n)
// 空间复杂度: O(m*n)
class Solution {
    private static final int[][] next = {{0, -1}, {0, 1}, {-1, 0}, {1, 0}};
    private int cnt = 0;
    private int m;
    private int n;
    private int k;
   
    public int movingCount(int m, int n, int k) {

        this.m = m;
        this.n = n;
        this.k = k;
        int cal;
        boolean[][] flag = new boolean[m][n];
        dfs(0,0,flag);
        return cnt;
    }
    private void dfs(int i, int j,boolean[][] flag ) {
     
        if (i < 0 || i >= m || j< 0 || j >= n || flag[i][j]) {
            return;
        }
        if(cal(i) + cal(j) > k) {
            return;
        }
        cnt++;
        flag[i][j] = true;
        for(int p = 0; p < 4; p++) {
            int newi= i + next[p][0];
            int newj= j + next[p][1];
            dfs(newi , newj , flag);
        }
    }
    private int cal(int num) {
        int ref = 0;
        while(num > 0) {
            ref += num % 10;
            num /= 10;
        }
        return ref;
    }
}


// TODO:

File: 剑指offer/13-机器人的运动范围P92/Solution.cpp
// Author : lihongyu
// Time   : 2019-10-26

class Solution {
public:
    int movingCount(int threshold, int rows, int cols)
    {
        bool* flag=new bool[rows*cols];
        for(int i=0;i<rows*cols;i++)
            flag[i]=false;
        int count=moving(threshold,rows,cols,0,0,flag);//从（0,0）坐标开始访问;
        delete[] flag;
        return count;
    }
    //计算最大移动位置
    int moving(int threshold,int rows,int cols,int i,int j,bool* flag)
    {
        int count=0;
        if(check(threshold,rows,cols,i,j,flag))
        {
            flag[i*cols+j]=true;
            //标记访问过，这个标志flag不需要回溯，因为只要被访问过即可。
           //因为如果能访问，访问过会加1.不能访问，也会标记下访问过。
            count=1+moving(threshold,rows,cols,i-1,j,flag)
                   +moving(threshold,rows,cols,i,j-1,flag)
                   +moving(threshold,rows,cols,i+1,j,flag)
                   +moving(threshold,rows,cols,i,j+1,flag);
        }
        return count;
    }
    //检查当前位置是否可以访问
    bool check(int threshold,int rows,int cols,int i,int j,bool* flag)
    {
        if(i>=0 && i<rows && j>=0 && j<cols
            && getSum(i)+getSum(j)<=threshold
            && flag[i*cols+j]==false)
           return true;
        return false;
    }
    //计算位置的数值
    int getSum(int number)
    {
        int sum=0;
        while(number>0)
        {
            sum+=number%10;
            number/=10;
        }
        return sum;
    }
};
